### DISCUSSION TITLE
Order by condition for data having same timestamp

### DISCUSSION THREAD

**Discussion by akshah1997** [finos/TimeBase-CE#213] [Q&A]
Hello,

We are trying to find the last message for each time interval (lets say per minute) for a particular stream.

Given that there are multiple messages having same timestamps, we want to select just one message based on some conditions and not based on when it was added to the stream.
Can you please guide/advise us on how to proceed?

Thanks.

**Comment by alex-karpovich** [comment]
Greetings, @akshah1997 

You can try to use QQL, we have special aggregates:    
https://kb.timebase.info/community/development/qql/QQL%205.5/qql-tut-functions#stateful-functions

useful functions: OVER time(interval)

or you can provide more details about your stream schema, so I can try to help with query

**Reply by akshah1997** [reply]
Thanks @alex-karpovich for reaching out. Greetings!

Considering the following sample data from _my_stream_ -

```
| timestamp               | symbol    | value   | priority |
| 2022-08-25 00:00:00.040 | abc       | 20      | 1        |
| 2022-08-25 00:00:00.666 | cde       | 10      | 2        |
| 2022-08-25 00:00:01.040 | efg       | 20      | 1        |
| 2022-08-25 00:00:01.291 | ghi       | 30      | 1        |
| 2022-08-25 00:00:01.600 | hij       | 203     | 2        |
| 2022-08-25 00:00:01.600 | abc       | 201     | 1        |
| 2022-08-25 00:00:01.600 | cde       | 202     | 3        |
```

I want to aggregate the data over every second and get the last value for each second.
I am using the following query for the same -

`select last{} (last(value)) as last_value_per_sec from my_stream over TIME (1s)`

The output would be -

```
| timestamp           | symbol    | value   | priority |
| 2022-08-25 00:00:00 | cde       | 10      | 2        |
| 2022-08-25 00:00:01 | cde       | 202     | 3        |
```

However, I want to get the output such that when multiple rows have same timestamps, I should be able to get row based on `priority` i.e. order by some condition.

The expected output - 

```
| timestamp           | symbol    | value   | priority |
| 2022-08-25 00:00:00 | cde       | 10      | 2        |
| 2022-08-25 00:00:01 | abc       | 201     | 1        |
```

Can you please help with the same?

Thanks a lot for your time.

**Reply by alex-karpovich** [reply]
I think the only way is to create a 'custom' aggregate function that can be used in QQL.
'custom' means implement some kind of java 'plugin' and use it with timebase server classpath
Let me check tomorrow if this is possible with Community Edition

**Reply by akshah1997** [reply]
Okay sure, thanks.

**Reply by akshah1997** [reply]
Hi @alex-karpovich, any updates please?
Thanks.

**Reply by ankitsoin** [reply]
So currently i have found easy workaround for same as in my case data is NOT real time.
 I have loaded the data in the specific way of priority i need it so when TimeBase picks the last record over 1 minute it gives same record what i needed.

But if I get to know how to use custom aggregator as @alex-karpovich mentioned above, i can use it for other complex logics which I may require while retrieving data.
Thanks

**Reply by alex-karpovich** [reply]
https://github.com/finos/TimeBase-CE/blob/main-6.1/java/timebase/computations-std/src/main/java/com/epam/deltix/computations/finanalysis/BollingerBands.java

here is sample of custom function, full description is here 

https://kb.timebase.info/community/development/qql/QQL%205.5/qql-tut-functions

you need to create own java library based on "computations-api" and add to the Timebase classpath
I will provide short instructions
